Sources/Core/ApplicationObserving.swift

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
+//
+// You are hereby granted a non-exclusive, worldwide, royalty-free license to use,
+// copy, modify, and distribute this software in source code or binary form for use
+// in connection with the web services and APIs provided by Facebook.
+//
+// As with any software that integrates with the Facebook platform, your use of
+// this software is subject to the Facebook Developer Principles and Policies
+// [http://developers.facebook.com/policy/]. This copyright notice shall be
+// included in all copies or substantial portions of the software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+// FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+// COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+// IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+import UIKit
+
+/**
+ Type-erasure wrapper for `ApplicationObserving` to allow for implementing `Equatable`
+ which helps to avoid storing and handling duplicate observers
+ */
+struct AnyApplicationObserving: Equatable {
+  let applicationObserving: ApplicationObserving
+
+  static func == (lhs: AnyApplicationObserving, rhs: AnyApplicationObserving) -> Bool {
+    return lhs.applicationObserving.isEqualTo(rhs.applicationObserving)
+  }
+}
+
+extension ApplicationObserving where Self: Equatable {
+  func isEqualTo(_ application: ApplicationObserving) -> Bool {
+    guard let otherApplication = application as? Self else {
+      return false
+    }
+    return self == otherApplication
+  }
+}
+
+protocol ApplicationObserving {
+  func isEqualTo(_ application: ApplicationObserving) -> Bool
+
+  func applicationDidBecomeActive(_ application: UIApplication)
```

Comment:
I think some functions could be optional, which I'm planning to apply for objc SDK.

Reply:
So the `optional` keyword can only be applied to an `@objc` protocol. The way to do this in Swift would be to have an extension where you provide a default implementation of the method. I haven't provided those here because I haven't run into a use case yet that requires it.

It's very easy to add those later as needed. :)